zcash_client_backend/src/sync.rs

```diff
@@ -0,0 +1,399 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::anyhow;
+use futures_util::TryStreamExt;
+use prost::Message;
+use tokio::{fs::File, io::AsyncWriteExt, task::JoinHandle};
+use tonic::{
+    body::BoxBody,
+    client::GrpcService,
+    codegen::{Body, Bytes, StdError},
+};
+use tracing::{debug, error, info};
+// TODO: Move `FsBlockDb` behind a `BlockCache` trait to break cyclic dependency.
+use zcash_client_sqlite::{chain::BlockMeta, FsBlockDb, FsBlockDbError};
+use zcash_primitives::{
+    consensus::{BlockHeight, Parameters},
+    merkle_tree::HashSer,
+};
+
+use crate::{
+    data_api::{
+        chain::{error::Error as ChainError, scan_cached_blocks, BlockSource, CommitmentTreeRoot},
+        scanning::{ScanPriority, ScanRange},
+        WalletCommitmentTrees, WalletRead, WalletWrite,
+    },
+    proto::service::{self, compact_tx_streamer_client::CompactTxStreamerClient},
+};
+
+pub(crate) fn get_block_path(fsblockdb_root: &Path, meta: &BlockMeta) -> PathBuf {
+    meta.block_file_path(&fsblockdb_root.join("blocks"))
+}
+
+/// Scans the chain until the wallet is up-to-date.
+pub async fn run<P, ChT, DbT>(
+    client: &mut CompactTxStreamerClient<ChT>,
+    params: &P,
+    fsblockdb_root: &Path,
+    db_cache: &mut FsBlockDb,
```

Comment:
Why couple this sync implementation to `FsBlockDb`? I can't use it because it's too slow. Can we use a trait for it instead?

Reply:
hi arnott, I am working on a trait here https://github.com/zcash/librustzcash/pull/1192

I also have a branch in my fork which updates this sync engine to build and run with the block cache trait so we dont need to use the SQLite or store blocks on disk

Reply:
Indeed, this is the TODO I noted at the very top of this PR: https://github.com/zcash/librustzcash/pull/1184#discussion_r1484594940